### PR TITLE
[quant][graph] Add valueQuantizable function

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36622 [quant][graph] Graph mode quantization support for sigmoid
* #36552 [quant][graph] Add quantized batch_norm2d_relu to graph mode
* #36549 [quant][graph] Add quantized batch_norm2d support to graph mode
* **#36548 [quant][graph] Add valueQuantizable function**

Summary:
Refactor to be able to observe based on inputs to ops

Test Plan:
test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21048963](https://our.internmc.facebook.com/intern/diff/D21048963)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
   }
 }
 
-// Returns true if the value is the weight to LSTM operator.
-bool isDynamicLSTMWeight(Value* v, Use use, bool is_dynamic) {
-  return is_dynamic && use.user->kind() == Symbol::aten("lstm") &&
-      (use.offset == 2);
+bool valueQuantizable(Use use, bool is_dynamic) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:989]: since this is checking the use, could you rename it to `useQuantizable`?

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
...
+  Node* n = use.user;
+  // static quantization
+  if (n->kind() == prim::CallFunction &&
+      getFuncName(n->inputs()[0]) == "batch_norm") {
+    return use.offset == 1;
+  }
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:995]: could you refactor this further to take a list of tuple of (name, offset) and check if the use is matching the name and offset?

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
...
+  // Dynamic quantization checks
+  if (is_dynamic) {
+    if (use.user->kind() == Symbol::aten("lstm")) {
+      return use.offset == 2;
+    }
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1000]: similar for this one

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:995]: some similar functions we already implemented: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization.cpp#L707, https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization.cpp#L713, please feel free to refactor these as well as you see fit

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -986,10 +986,20 @@ void InsertObserversHelper::preprocess(
...
+  }
+  return nodeQuantizable(n);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1002]: I feel hiding this here is not very explicit, I think moving this outside and use `nodeQuantizable && useQuantizable` is probably better

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp]: I've added the list of tuple in the function itself. I didn't want to spend too much time refactoring the other code, so I used it in the best way I saw fit, given that the conditions we were returning on in both cases was different.

File: torch/csrc/jit/passes/quantization.cpp [hunk 5]
```diff
@@ -986,10 +987,25 @@ void InsertObserversHelper::preprocess(
...
+      return v == n->inputs().at(func_arg.arg_index);
+    }
+  }
+
+  for (const auto& func_arg : call_func_args) {
+    if (n->kind() == prim::CallFunction &&
+        getFuncName(n->inputs()[0]) == func_arg.func_name) {
+      return v == n->inputs().at(func_arg.arg_index);
+    }
+  }
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1006]: I think you can use isAtenFuncNthArg and isCallFunctionNthArg here, also you can refactor these functions to not take value as input, and just compare `use.offset` with the `arg_index`

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:1006]: the logic seems a bit different here actually, and this is due to the convoluted check logic that we fall back to nodeQuantizable when the check fails. Can we add a default index for aten and call function in nodes in nodeQuantizable?

File: torch/csrc/jit/passes/quantization.cpp [hunk 6]
```diff
@@ -675,24 +676,28 @@ graph(%self, %a, %b, %inplace):
 };
 
 // Check if `use` is an aten function of name `func_name` and if value
-// `v` is the nth argument of the function
-bool isAtenFuncNthArg(
-    Value* v,
+// `v` is the nth argument (if provided) of the function.
+bool matchAtenFuncToUse(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:680]: Can we remove `Node*` and `Value` and just have `Use` here?

File: torch/csrc/jit/passes/quantization.cpp [hunk 7]
```diff
@@ -675,24 +676,28 @@ graph(%self, %a, %b, %inplace):
...
+      (n.has_value() && v.has_value() ? v.value() == use->inputs().at(n.value())
+                                      : true);
 }
 
 // Check if `use` is a CallFunction of name `func_name` and if value
-// `v` is the nth argument of the function
-bool isCallFunctionNthArg(
-    Value* v,
+// `v` is the nth argument (if provided) of the function
+bool matchCallFuncToUse(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:692]: same for this one

File: torch/csrc/jit/passes/quantization.cpp [hunk 8]
```diff
@@ -986,10 +992,30 @@ void InsertObserversHelper::preprocess(
...
+  Node* n = use.user;
+  bool result = nodeQuantizable(n);
+  if (!result) {
+    return false;
+  }
+  // Special checks for ops that do not require observers for all input tensors.
+  // For each operator in this list observers are inserted for the input based
+  // on the index specified.
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({{"lstm", 2}});
+  const CallFuncArgs& call_func_args = CallFuncArgs({{"batch_norm", 1}});
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1005]: this change belongs to next PR

File: torch/csrc/jit/passes/quantization.cpp (hunk 6 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:680]: Isn't the value necessary to compare this `v == use->inputs().at(n)`?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:680]: no, you can do use.offset == n

File: torch/csrc/jit/passes/quantization.cpp [hunk 9]
```diff
@@ -675,24 +676,26 @@ graph(%self, %a, %b, %inplace):
...
+// `v` is the nth argument (if provided) of the function
+bool matchCallFuncToUse(
+    const Use& use,
     const std::string& func_name,
-    int n) {
-  return use->kind() == prim::CallFunction &&
-      getFuncName(use->inputs()[0]) == func_name && v == use->inputs().at(n);
+    c10::optional<int> n) {
+  Node* node = use.user;
+  return node->kind() == prim::CallFunction &&
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:696]: false --> true?

File: torch/csrc/jit/passes/quantization.cpp [hunk 10]
```diff
@@ -986,10 +988,29 @@ void InsertObserversHelper::preprocess(
...
-      (use.offset == 2);
+bool useQuantizable(const Use& use, bool is_dynamic) {
+  bool result = nodeQuantizable(use.user);
+  if (!result) {
+    return false;
+  }
+  // Special checks for ops that do not require observers for all input tensors.
+  // For each operator in this list observers are inserted for the input based
+  // on the index specified.
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({{"lstm", 2}});
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:999]: we also need to check `is_dynamic` for "lstm" node..

File: torch/csrc/jit/passes/quantization.cpp [hunk 11]
```diff
@@ -986,10 +988,29 @@ void InsertObserversHelper::preprocess(
...
 
-// Returns true if the value is the weight to LSTM operator.
-bool isDynamicLSTMWeight(Value* v, Use use, bool is_dynamic) {
-  return is_dynamic && use.user->kind() == Symbol::aten("lstm") &&
-      (use.offset == 2);
+bool useQuantizable(const Use& use, bool is_dynamic) {
+  bool result = nodeQuantizable(use.user);
+  if (!result) {
+    return false;
+  }
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:995]: move this to the end of the function?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:995]: - just return nodeQUantizable(use.user) in the end

File: torch/csrc/jit/passes/quantization.cpp (hunk 10 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:999]: I think that won't be necessary because for dynamic quant we insert observers at the input. Which should be same as static quant. We have a check for `is_dynamic` to insert observer at the output of node in the parent function.

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:999]: OK, then do you need `is_dynamic` as argument here?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:999]: we don't support static quant right now though

File: torch/csrc/jit/passes/quantization.cpp [hunk 12]
```diff
@@ -986,10 +988,31 @@ void InsertObserversHelper::preprocess(
...
+
+  for (const auto& func_arg : call_func_args) {
+    if (matchCallFuncToUse(use, func_arg.func_name, c10::nullopt)) {
+      return use.offset == func_arg.arg_index;
+    }
+  }
+  // Dynamic quantized ops that require special handling for inputs.
+  if (is_dynamic) {
+    if (matchAtenFuncToUse(use, "lstm", c10::nullopt)) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:1010]: nit: merge the conditions

File: torch/csrc/jit/passes/quantization.cpp [hunk 13]
```diff
@@ -986,10 +988,31 @@ void InsertObserversHelper::preprocess(
...
-// Returns true if the value is the weight to LSTM operator.
-bool isDynamicLSTMWeight(Value* v, Use use, bool is_dynamic) {
-  return is_dynamic && use.user->kind() == Symbol::aten("lstm") &&
-      (use.offset == 2);
+bool useQuantizable(const Use& use, bool is_dynamic) {
+  // Special checks for ops that do not require observers for all input tensors.
+  // For each operator in this list observers are inserted for the input based
+  // on the index specified.
+  const AtenFuncArgs& aten_func_args = AtenFuncArgs({});
+  const CallFuncArgs& call_func_args = CallFuncArgs({});
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:996]: could you expose these two lists in the beginning of the file so that it's easier to find and change?

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks! left a few more comments, please address them before landing

File: torch/csrc/jit/passes/quantization.cpp [hunk 14]
```diff
@@ -675,24 +676,26 @@ graph(%self, %a, %b, %inplace):
...
+    const Use& use,
     const std::string& func_name,
-    int n) {
-  return use->kind() == prim::CallFunction &&
-      getFuncName(use->inputs()[0]) == func_name && v == use->inputs().at(n);
+    c10::optional<int> n) {
+  Node* node = use.user;
+  return node->kind() == prim::CallFunction &&
+      getFuncName(node->inputs()[0]) == func_name &&
+      (n.has_value() ? (n.value() == use.offset) : true);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:698]: actually this can be changed to `!n.has_value() || n.value() == use.offset`

File: torch/csrc/jit/passes/quantization.cpp (hunk 13 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp:996]: Also, a follow up here is, can we remove the isBias check for conv and linear and put the special case in the list?

Comment (supriyar) [reply torch/csrc/jit/passes/quantization.cpp:996]: > Also, a follow up here is, can we remove the isBias check for conv and linear and put the special case in the list?

I think we can do this separately, I also see `isWeight` check that uses these lists.

File: torch/csrc/jit/passes/quantization.cpp [hunk 15]
```diff
@@ -110,6 +111,20 @@ std::vector<std::string> _single_input_general_aten_funcs = {
...
+};
+
+using AtenFuncArgs = std::vector<FuncArg>;
+using CallFuncArgs = std::vector<FuncArg>;
+
+// Special checks for ops that do not require observers for all input tensors.
+// For each operator in this list observers are inserted for the input based
+// on the index specified.
+AtenFuncArgs _aten_func_args = {};
+CallFuncArgs _call_func_args = {};
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:126]: can you use better names for these?
